[Screen Time Refactoring] Ensure donations do not occur in private browsing / ephemeral sessions

jesxilin opened the pull request:
#### 577fb0fa5f923e6c550e5d557380e7e3ed663a70
<pre>
[Screen Time Refactoring] Ensure donations do not occur in private browsing / ephemeral sessions
<a href="https://bugs.webkit.org/show_bug.cgi?id=286493">https://bugs.webkit.org/show_bug.cgi?id=286493</a>
<a href="https://rdar.apple.com/140439254">rdar://140439254</a>

Reviewed by Aditya Keerthi and Wenson Hsieh.

Set suppressUsageRecording on a STWebpageController to ensure
screen time donations are not accounted during
private browsing or ephemeral sessions.

Added API tests to ensure flag is being set properly.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageController]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(testSuppressUsageRecordingWithDataStore):
(TEST(ScreenTime, PersistentSession)):
(TEST(ScreenTime, NonPersistentSession)):

Canonical link: <a href="https://commits.webkit.org/289370@main">https://commits.webkit.org/289370@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8efa0b667f4d38f055dac4a89b0a2a896514336a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86717 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6224 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41052 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/91562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/37449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88766 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6492 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14282 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/91562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/37449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89720 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/78511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/91562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32848 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/36567 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33733 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93457 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13870 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/10058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/75845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14071 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74339 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75034 "Passed tests") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/19349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17753 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13477 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13893 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/19153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/13631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/17076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/15416 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -369,6 +369,8 @@ - (void)_installScreenTimeWebpageController
         if ([_screenTimeWebpageController respondsToSelector:@selector(setProfileIdentifier:)])
             [_screenTimeWebpageController setProfileIdentifier:[_configuration websiteDataStore].identifier.UUIDString];
 
+        [_screenTimeWebpageController setSuppressUsageRecording: ![_configuration websiteDataStore].isPersistent];
```
No space here
```suggestion
        [_screenTimeWebpageController setSuppressUsageRecording:![_configuration websiteDataStore].isPersistent];
```

rr-codes replied:
nit: no space after the :

whsieh replied:
```suggestion
        [_screenTimeWebpageController setSuppressUsageRecording:![_configuration websiteDataStore].isPersistent];
```

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -61,6 +61,7 @@ @interface STWebpageController ()
 
 @interface STWebpageController (Staging_138865295)
 @property (nonatomic, copy) NSString *profileIdentifier;
+@property (setter=setSuppressUsageRecording:) BOOL suppressUsageRecording;
```
Is this forward declaration necessary? (Since this is existing public API)

If it needs to be forward declared, then I would recommend moving it out of this category (`Staging_138865295`) and into a separate category with the appropriate radar ID.

jesxilin replied:
It was not necessary... removing now!

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -220,4 +221,61 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+        })
+    };
+
+    // A non-persistent data store means data will not be persistent.
```
I don't think this comment is needed

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -220,4 +221,61 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
     EXPECT_WK_STREQ(identifier, uuidString.get());
 }
+
+TEST(ScreenTime, EphemeralSessionOff)
```
maybe instead of EphemeralSessionOff/On, it can be PersistentSession/EphemeralSession?

jesxilin replied:
As in PersistentSession vs NonPersistentSession?

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -220,4 +221,61 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+        PAL::getSTWebpageControllerClass(),
+        @selector(setSuppressUsageRecording:),
+        imp_implementationWithBlock(^(id object, bool ephemeralSession) {
+            suppressUsageRecording = ephemeralSession;
```
Nit, since the current variable name doesn't really make sense in the context of the Screen Time method:

```suggestion
        imp_implementationWithBlock(^(id object, bool value) {
            suppressUsageRecording = value;
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -220,4 +221,61 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+        PAL::getSTWebpageControllerClass(),
+        @selector(setSuppressUsageRecording:),
+        imp_implementationWithBlock(^(id object, bool ephemeralSession) {
+            suppressUsageRecording = ephemeralSession;
```
```suggestion
        imp_implementationWithBlock(^(id object, bool value) {
            suppressUsageRecording = value;
```

aprotyas reviewed (commented):
You could imagine sharing most of the code between the tests. You'd want to pass in a `WKWebsiteDataStore` instance and an expected value for `suppressUsageRecording`.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -220,4 +221,61 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+        })
+    };
+
+    // A default data store has data as persistent.
```
This comment can also be removed, since it's implied by the test.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -55,6 +55,34 @@
+    if (suppressUsageRecordingExpectation)
+        EXPECT_TRUE(suppressUsageRecording);
+    else
+        EXPECT_FALSE(suppressUsageRecording);
```
Could use `EXPECT_EQ`.

jesxilin replied:
Ah, that's better, I'll change!

jesxilin committed:
[Screen Time Refactoring] Ensure donations do not occur in private browsing / ephemeral sessions
https://bugs.webkit.org/show_bug.cgi?id=286493
rdar://140439254

Reviewed by Aditya Keerthi and Wenson Hsieh.

Set suppressUsageRecording on a STWebpageController to ensure
screen time donations are not accounted during
private browsing or ephemeral sessions.

Added API tests to ensure flag is being set properly.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageController]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(testSuppressUsageRecordingWithDataStore):
(TEST(ScreenTime, PersistentSession)):
(TEST(ScreenTime, NonPersistentSession)):

Canonical link: https://commits.webkit.org/289370@main
